core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -776,7 +801,13 @@ private[log] class Cleaner(val id: Int,
        *   2) The message doesn't has value but it can't be deleted now.
        */
       val latestOffsetForKey = record.offset() >= foundOffset
-      val isRetainedValue = record.hasValue || retainDeletes
+      val supportDeleteHorizon = batch.magic() >= RecordBatch.MAGIC_VALUE_V2
+      val shouldRetainDeletes =
```

Comment:
nit: maybe turn this into a `def` since we don't even use the computed value unless `record.hasValue` is false.